File: app/src/main/java/uk/co/sundroid/activity/data/fragments/DayDetailPlanetsFragment.java
package uk.co.sundroid.activity.data.fragments;

import android.os.Handler;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import uk.co.sundroid.R;
import uk.co.sundroid.util.astro.Body;
import uk.co.sundroid.util.astro.BodyDay;
import uk.co.sundroid.util.astro.RiseSetType;
import uk.co.sundroid.util.astro.math.BodyPositionCalculator;
import uk.co.sundroid.domain.LocationDetails;
import uk.co.sundroid.util.theme.ThemePalette;
import uk.co.sundroid.util.time.TimeHelper;
import uk.co.sundroid.util.time.TimeHelper.Time;
import uk.co.sundroid.util.geometry.BearingHelper;

import java.util.*;

public class DayDetailPlanetsFragment extends AbstractDayFragment {

    private final Handler handler = new Handler();

	@Override
	protected int getLayout() {
		return R.layout.frag_data_daydetail_planets;
	}

	protected void update(final LocationDetails location, final Calendar calendar, final View view) throws Exception {

        Thread thread = new Thread() {
            @Override
            public void run() {
                if (!isSafe()) {
                    return;
                }

                final Map<Body, BodyDay> planetDays = new LinkedHashMap<>();
                for (Body body : Body.values()) {
                    if (body != Body.SUN && body != Body.MOON) {
                        planetDays.put(body, BodyPositionCalculator.calcDay(body, location.getLocation(), calendar, true));
                    }
                }

                handler.post(() -> {
                    if (!isSafe()) {
                        return;
                    }
                    ViewGroup planetsDataBox = view.findViewById(R.id.planetsDataBox);
                    for (Map.Entry<Body, BodyDay> planetEntry : planetDays.entrySet()) {
                        Body planet = planetEntry.getKey();
                        BodyDay planetDay = planetEntry.getValue();
                        View planetRow = getActivity().getLayoutInflater().inflate(R.layout.frag_data_daydetail_planets_planet, planetsDataBox);

                        textInView(planetRow, R.id.planetName, planet.name());

                        boolean noTransit = false;
                        boolean noUptime = false;

                        if (planetDay.getRiseSetType() != RiseSetType.SET && planetDay.getTransitAppElevation() > 0) {
                            Time noon = TimeHelper.formatTime(getApplicationContext(), planetDay.getTransit(), false);
                            showInView(planetRow, R.id.planetTransit);
                            showInView(planetRow, R.id.planetTransitTime, noon.time + noon.marker + "  " + BearingHelper.formatElevation(planetDay.getTransitAppElevation()));
                        } else {
                            removeInView(planetRow, R.id.planetTransit);
                            noTransit = true;
                        }

                        if (planetDay.getRiseSetType() == RiseSetType.RISEN || planetDay.getRiseSetType() == RiseSetType.SET) {
                            showInView(planetRow, R.id.planetSpecial, planetDay.getRiseSetType() == RiseSetType.RISEN ? "Risen all day" : "Set all day");
                            removeInView(planetRow, R.id.planetEvtsRow, R.id.planetEvt1, R.id.planetEvt2, R.id.planetUptime);
                            noUptime = true;
                        } else {
                            removeInView(planetRow, R.id.planetSpecial);
                            removeInView(planetRow, R.id.planetEvt1, R.id.planetEvt2);
                            showInView(planetRow, R.id.planetEvtsRow);
                            Set<SummaryEvent> events = new TreeSet<>();
                            if (planetDay.getRise() != null) {
                                events.add(new SummaryEvent("Rise", planetDay.getRise(), planetDay.getRiseAzimuth()));
                            }
                            if (planetDay.getSet() != null) {
                                events.add(new SummaryEvent("Set", planetDay.getSet(), planetDay.getSetAzimuth()));
                            }
                            int index = 1;
                            for (SummaryEvent event : events) {
                                int rowId = view("planetEvt" + index);
                                int timeId = view("planetEvt" + index + "Time");
                                int azId = view("planetEvt" + index + "Az");
                                int imgId = view("planetEvt" + index + "Img");

                                Time time = TimeHelper.formatTime(getApplicationContext(), event.getTime(), false);
                                String az = BearingHelper.formatBearing(getApplicationContext(), event.getAzimuth(), location.getLocation(), event.getTime());

                                textInView(planetRow, timeId, time.time + time.marker);
                                textInView(planetRow, azId, az);
                                showInView(planetRow, rowId);

                                if (event.getName().equals("Rise")) {
                                    ((ImageView)planetRow.findViewById(imgId)).setImageResource(ThemePalette.getRiseArrow());
                                } else {
                                    ((ImageView)planetRow.findViewById(imgId)).setImageResource(ThemePalette.getSetArrow());
                                }

                                index++;
                            }

                            if (planetDay.getUptimeHours() > 0 && planetDay.getUptimeHours() < 24) {
                                showInView(planetRow, R.id.planetUptime);
                                showInView(planetRow, R.id.planetUptimeTime, TimeHelper.formatDurationHMS(getApplicationContext(), planetDay.getUptimeHours(), false));
                            } else {
                                removeInView(planetRow, R.id.planetUptime);
                            }

                        }

                        if (noTransit && noUptime) {
                            removeInView(planetRow, R.id.planetTransitUptime);
                        } else {
                            showInView(planetRow, R.id.planetTransitUptime);
                        }

                    }
                    showInView(view, R.id.planetsDataBox);
                });
            }
        };
        thread.start();
    }

}

File: app/src/main/kotlin/uk/co/sundroid/activity/data/fragments/DayDetailPlanetsFragment.kt
package uk.co.sundroid.activity.data.fragments

import android.view.View
import android.view.ViewGroup
import uk.co.sundroid.R
import uk.co.sundroid.R.id.*
import uk.co.sundroid.util.astro.Body
import uk.co.sundroid.util.astro.BodyDay
import uk.co.sundroid.util.astro.RiseSetType
import uk.co.sundroid.util.astro.math.BodyPositionCalculator
import uk.co.sundroid.util.async.async
import uk.co.sundroid.util.geometry.formatBearing
import uk.co.sundroid.util.geometry.formatElevation
import uk.co.sundroid.util.theme.getRiseArrow
import uk.co.sundroid.util.theme.getSetArrow
import uk.co.sundroid.util.time.formatDurationHMS
import uk.co.sundroid.util.html
import uk.co.sundroid.util.time.formatTimeStr
import java.util.*

class DayDetailPlanetsFragment : AbstractDayDetailFragment() {

    override val layout: Int
        get() = R.layout.frag_data_daydetail_planets

    override fun updateData(view: View) {
        val location = getLocation()
        val calendar = getDateCalendar()

        async(
                inBackground = { Body.PLANETS.associate { Pair(it, BodyPositionCalculator.calcDay(it, location.location, calendar, true)) } },
                onDone = { days: Map<Body, BodyDay> ->
                    if (isSafe) {
                        val wrapper = view.findViewById<ViewGroup>(planetsDataBox)
                        wrapper.removeAllViews()
                        for ((planet, day) in days) {
                            val row = inflate(R.layout.frag_data_daydetail_planets_planet)

                            text(row, planetName, planet.name)

                            var noTransit = false
                            var noUptime = false

                            if (day.riseSetType !== RiseSetType.SET && day.transitAppElevation > 0) {
                                val noon = formatTimeStr(requireContext(), day.transit!!, false, html = true)
                                show(row, planetTransit)
                                show(row, planetTransitTime, html("$noon &nbsp; ${formatElevation(day.transitAppElevation)}"))
                            } else {
                                remove(row, planetTransit)
                                noTransit = true
                            }

                            if (day.riseSetType === RiseSetType.RISEN || day.riseSetType === RiseSetType.SET) {
                                show(row, planetSpecial, if (day.riseSetType === RiseSetType.RISEN) "Risen all day" else "Set all day")
                                remove(row, planetEvtsRow, planetEvt0, planetEvt1, planetUptime)
                                noUptime = true
                            } else {
                                remove(row, planetSpecial)
                                remove(row, planetEvt0, planetEvt1)
                                show(row, planetEvtsRow)
                                val events = TreeSet<RiseSetEvent>()
                                day.rise?.let { events.add(RiseSetEvent("Rise", it, day.riseAzimuth)) }
                                day.set?.let { events.add(RiseSetEvent("Set", it, day.setAzimuth)) }
                                events.forEachIndexed { index, event ->
                                    val rowId = view("planetEvt$index")
                                    val timeId = view("planetEvt${index}Time")
                                    val azId = view("planetEvt${index}Az")
                                    val imgId = view("planetEvt${index}Img")

                                    val time = formatTimeStr(requireContext(), event.time, false, html = true)
                                    val az = formatBearing(requireContext(), event.azimuth, location.location, event.time)

                                    text(row, timeId, html(time))
                                    text(row, azId, az)
                                    show(row, rowId)
                                    image(row, imgId, if (event.name == "Rise") getRiseArrow() else getSetArrow())
                                }

                                if (day.uptimeHours > 0 && day.uptimeHours < 24) {
                                    show(row, planetUptime)
                                    show(row, planetUptimeTime, html(formatDurationHMS(requireContext(), day.uptimeHours, false, html = true)))
                                } else {
                                    remove(row, planetUptime)
                                }
                            }

                            if (noTransit && noUptime) {
                                remove(row, planetTransitUptime)
                            } else {
                                show(row, planetTransitUptime)
                            }
                            wrapper.addView(row)

                        }
                        show(wrapper)
                    }
                }
        )
    }

}